x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,230 @@
+	}
+
+	return slots
+}
+
+// group the slots
+func GroupSlots(slots []*PairingSlot) []PairingSlotGroup {
+	slotGroups := []PairingSlotGroup{}
+	if len(slots) == 0 {
+		utils.LavaFormatError("no slots", sdkerrors.ErrLogic)
+		return slotGroups
+	}
+
+	for k := range slots {
+		foundGroup := false
+		for i := range slotGroups {
+			diff := slots[k].Diff(slotGroups[i].Slot)
+			if len(diff.Reqs) == 0 {
+				slotGroups[i].Count += 1
+				foundGroup = true
+				break
+			}
+		}
+
+		if !foundGroup {
+			newGroup := NewPairingSlotGroup(slots[k])
+			slotGroups = append(slotGroups, *newGroup)
+		}
+	}
+
+	return slotGroups
+}
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func GetStrategy() ScoreStrategy {
+	return uniformStrategy
+}
+
+// calculates the final pairing score for all slot groups (with strategy)
+// we calculate only the diff between the current and previous slot groups
+func CalcPairingScore(scores []*PairingScore, strategy ScoreStrategy, diffSlot *PairingSlot, minStake sdk.Int) error {
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight, ok := strategy[reqName]
+		if !ok {
+			return utils.LavaFormatError("req not in strategy", sdkerrors.ErrKeyNotFound,
+				utils.Attribute{Key: "req_name", Value: reqName})
+		}
+
+		for _, score := range scores {
+			// normalize stake so we won't overflow the score result (uint64)
+			newScoreComp := req.Score(*score.Provider, weight)
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				if prevReqScoreComp == 0 {
+					return utils.LavaFormatError("previous req score is zero", fmt.Errorf("invalid req score"),
+						utils.Attribute{Key: "req_name", Value: reqName})
+				}
+				score.Score /= prevReqScoreComp
+			}
+			score.Score *= newScoreComp
+
+			// update the score component map
+			score.ScoreComponents[reqName] = newScoreComp
+		}
+	}
+
+	return nil
+}
+
+// given a list of scores, pick a <group-count> providers with a pseudo-random weighted choice
+func PickProviders(ctx sdk.Context, projectIndex string, scores []*PairingScore, groupCount uint64, block uint64, chainID string, epochHash []byte, indexToSkipPtr *map[int]bool) (returnedProviders []epochstoragetypes.StakeEntry) {
+	scoreSum := sdk.NewUint(0)
+	hashData := make([]byte, 0)
+	for _, providerScore := range scores {
+		scoreSum = scoreSum.Add(sdk.NewUint(providerScore.Score))
+	}
+	if scoreSum.IsZero() {
+		// list is empty
+		return returnedProviders
+	}
+
+	// add the session start block hash to the function to make it as unpredictable as we can
+	hashData = append(hashData, epochHash...)
+	hashData = append(hashData, chainID...)      // to make this pairing unique per chainID
+	hashData = append(hashData, projectIndex...) // to make this pairing unique per consumer
+
+	indexToSkip := *indexToSkipPtr
```

Comment:
I see how it's used, but unclear why it comes from the caller.
(why is the arg a pointer to begin with?)

Reply:
to save time (as discussed in the daily), I'm keeping this code.